### PR TITLE
ssl: allow openssl engine support to be optional

### PR DESCRIPTION
**Pull request by ewertons**
This is a port from https://github.com/Azure/azure-c-shared-utility/pull/596

### CODE REVIEW

File: adapters/x509_openssl.c [hunk 1]
```diff
@@ -139,8 +141,21 @@ static int load_key_RSA(SSL_CTX* ssl_ctx, EVP_PKEY* evp_key)
...
     /*Codes_SRS_X509_OPENSSL_02_005: [ x509_openssl_add_credentials shall use PEM_read_bio_RSAPrivateKey to read the x509 private key. ]*/
-    RSA* privatekey = EVP_PKEY_get1_RSA(evp_key);
-    if (privatekey == NULL)
+    RSA *privatekey;
+#ifdef defined(OPENSSL_VERSION_NUMBER) AND (OPENSSL_VERSION_NUMBER >= OPENSSL_VERSION_3)
+    if (evp_key->type == EVP_PKEY_RSA || evp_key->type == EVP_PKEY_RSA_PSS)
+    {
+        privatekey = NULL;
+
+        if (privatekey != NULL)
```
Comment (ericwol-msft) [inline adapters/x509_openssl.c:150]: what does this do? looks like privatekey is null.

Comment (ewertons) [reply adapters/x509_openssl.c]: Sorry, I'm making changes to the file, I'll ping you once it's ready for review.

File: CMakeLists.txt [hunk 2]
```diff
@@ -115,6 +120,11 @@ if(${use_openssl})
     if (NOT TARGET OpenSSL::SSL OR NOT TARGET OpenSSL::Crypto OR NOT ${OPENSSL_INCLUDE_DIR})
         find_package(OpenSSL REQUIRED)
     endif()
+
+    if (DEFINED OPENSSL_VERSION AND (${OPENSSL_VERSION} GREATER_EQUAL 3))
```
Comment (CIPop) [inline CMakeLists.txt:124]: Nit: worth adding a short comment with explanation of support for OpenSSL v1 using the v3 compat layer (would be nice to also have a link to official documentation).

File: CMakeLists.txt [hunk 3]
```diff
@@ -29,6 +29,7 @@ option(use_builtin_httpapi "set use_builtin_httpapi to ON to use the built-in ht
 option(use_cppunittest "set use_cppunittest to ON to build CppUnitTest tests on Windows (default is OFF)" OFF)
 option(suppress_header_searches "do not try to find headers - used when compiler check will fail" OFF)
 option(use_custom_heap "use externally defined heap functions instead of the malloc family" OFF)
+option(no_openssl_engine "Disables the use of ENGINEs in OpenSSL, if the target version supports it" OFF)
```
Comment (CIPop) [inline CMakeLists.txt:32]: Should the last part of the sentence be removed? (Disabling the use doesn't depend on target version support.)

**Review by CIPop — APPROVED** [review]
Thanks @ewertons !

File: CMakeLists.txt (hunk 3 above)
Comment (ewertons) [reply CMakeLists.txt:32]: That's true. Removed.

File: CMakeLists.txt (hunk 2 above)
Comment (ewertons) [reply CMakeLists.txt:124]: Added.

    # The block below enables the v1 back-compatibility layer in OpenSSL 3,
    # if using that version or later. For reference, please check the OpenSSL
    # official documentation: https://www.openssl.org/docs/man3.0/man7/openssl_user_macros.html

**Pull request merged**